addons/delivery/models/stock_picking.py

```diff
@@ -126,9 +137,23 @@ def _send_confirmation_email(self):
         for pick in self:
             if pick.carrier_id:
                 if pick.carrier_id.integration_level == 'rate_and_ship' and pick.picking_type_code != 'incoming':
-                    pick.send_to_shipper()
+                    if pick.weight_at_label != 0 and pick.weight_at_label != pick.shipping_weight and pick.carrier_tracking_ref:
+                        pick.activity_schedule(
+                        'mail.mail_activity_todo',
+                        user_id=pick.user_id.id if pick.user_id else self.env.user.id,
+                        summary='Re-generate the shipping label')
+                    if not pick.carrier_tracking_ref and pick.picking_type_id.print_label:
+                        pick.send_to_shipper()
         return super(StockPicking, self)._send_confirmation_email()
 
+    def action_print_label(self):
+        for pick in self.filtered(lambda x: x.carrier_id and x.carrier_id.integration_level == 'rate_and_ship' and x.picking_type_code != 'incoming'):
+            if pick.move_line_ids.filtered(lambda l: not l.qty_done):
```

Comment:
use `any()` instead. It's quicker than `filtered` in that case as `any` will stop looping through the recordset and return `False` in case of non-processed move line